dao-erc721/src/erc721.rs

```diff
@@ -0,0 +1,97 @@
+    Address,
+};
+use casper_types::{bytesrepr::Bytes, U256};
+
+use crate::{
+    core::ERC721Token,
+    extensions::{BurnableERC721, MetadataERC721, MintableERC721},
+};
+
+use delegate::delegate;
+
+pub type TokenId = U256;
+pub type TokenUri = String;
+
+#[casper_contract_interface]
+pub trait ERC721Interface {
+    fn init(&mut self, name: String, symbol: String);
+    fn name(&self) -> String;
+    fn symbol(&self) -> String;
+    fn owner_of(&self, token_id: TokenId) -> Option<Address>;
+    fn balance_of(&self, owner: Address) -> U256;
+    fn total_supply(&self) -> U256;
+    fn token_uri(&self, token_id: TokenId) -> TokenUri;
+    fn base_uri(&self) -> TokenUri;
+    fn approve(&mut self, to: Option<Address>, token_id: TokenId);
+    fn get_approved(&self, token_id: TokenId) -> Option<Address>;
+    fn set_approval_for_all(&mut self, operator: Address, approved: bool);
+    fn is_approved_for_all(&self, owner: Address, operator: Address) -> bool;
+    fn transfer_from(&mut self, owner: Address, recipient: Address, token_id: TokenId);
+    fn safe_transfer_from(&mut self, owner: Address, recipient: Address, token_id: TokenId);
+    fn safe_transfer_from_with_data(
+        &mut self,
+        owner: Address,
+        recipient: Address,
+        token_id: TokenId,
+        data: Bytes,
+    );
+    fn mint(&mut self, to: Address, token_id: TokenId);
+    fn burn(&mut self, token_id: TokenId);
+}
+
+#[derive(Instance)]
+pub struct ERC721 {
+    core: ERC721Token,
+    metadata: MetadataERC721,
+}
+
+impl ERC721Interface for ERC721 {
+    delegate! {
+        to self.metadata {
+            fn init(&mut self, name: String, symbol: String);
+            fn name(&self) -> String;
+            fn symbol(&self) -> String;
+        }
+
+        to self.core {
+            fn owner_of(&self, token_id: TokenId) -> Option<Address>;
+            fn balance_of(&self, owner: Address) -> U256;
+            fn total_supply(&self) -> U256;
+            fn token_uri(&self, token_id: TokenId) -> TokenUri;
+            fn base_uri(&self) -> TokenUri;
+            fn approve(&mut self, to: Option<Address>, token_id: TokenId);
+            fn get_approved(&self, token_id: TokenId) -> Option<Address>;
+            fn set_approval_for_all(&mut self, operator: Address, approved: bool);
+            fn is_approved_for_all(&self, owner: Address, operator: Address) -> bool;
+            fn transfer_from(&mut self, owner: Address, recipient: Address, token_id: TokenId);
+        }
+    }
+
+    fn safe_transfer_from(&mut self, owner: Address, recipient: Address, token_id: TokenId) {
+        self.core
+            .safe_transfer_from(owner, recipient, token_id, None);
+    }
+
+    fn safe_transfer_from_with_data(
+        &mut self,
+        owner: Address,
+        recipient: Address,
+        token_id: TokenId,
+        data: Bytes,
+    ) {
+        self.core
+            .safe_transfer_from(owner, recipient, token_id, Some(data));
+    }
```

Comment:
Shouldn't we merge those two with Option<data>? It will be closer to the standard

Reply:
Hmm, why not. I thought having two functions will be closer to the standard, and you could write less code by not passing `None` argument.